File: 0152-maximum-product-subarray/0152-maximum-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function (nums) {
    let minimumProduct = nums[0];
    let maximumProduct = nums[0];
    let ans = nums[0];
    //used for temporary
    let x, y;

    for (let num of nums.slice(1)) {
        x = Math.max(num, maximumProduct * num, minimumProduct * num);
        y = Math.min(num, minimumProduct * num, maximumProduct * num);
        [maximumProduct, minimumProduct] = [x, y]
        ans = Math.max(ans, maximumProduct);
    }

    return ans;
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        prev_min = prev_max = glob_max = nums[0]
        for i in nums[1:]:
            curr_min = min(prev_min*i, prev_max*i, i)
            curr_max = max(prev_max*i, prev_min*i, i)
            glob_max = max(glob_max, curr_max)
            prev_min = curr_min
            prev_max = curr_max
        return glob_max